x/pairing/keeper/scores/score.go

```diff
@@ -0,0 +1,230 @@
+	// init slot array (should be as the number of providers to pair)
+	slots := make([]*PairingSlot, policy.MaxProvidersToPair)
+
+	// all slots should consider the stake, so we init them with stakeReq
+	stakeReq := StakeReq{MinStake: minStake}
+	slotReqs := map[string]ScoreReq{stakeReq.GetName(): stakeReq}
+	for i := range slots {
+		slots[i] = NewPairingSlot(slotReqs)
+	}
+
+	return slots
+}
+
+// group the slots
+func GroupSlots(slots []*PairingSlot) []PairingSlotGroup {
+	slotGroups := []PairingSlotGroup{}
+	if len(slots) == 0 {
+		utils.LavaFormatError("no slots", sdkerrors.ErrLogic)
+		return slotGroups
+	}
+
+	for k := range slots {
+		foundGroup := false
+		for i := range slotGroups {
+			diff := slots[k].Diff(slotGroups[i].Slot)
+			if len(diff.Reqs) == 0 {
+				slotGroups[i].Count += 1
+				foundGroup = true
+				break
+			}
+		}
+
+		if !foundGroup {
+			newGroup := NewPairingSlotGroup(slots[k])
+			slotGroups = append(slotGroups, *newGroup)
+		}
+	}
+
+	return slotGroups
+}
+
+// TODO: currently we'll use weight=1 for all reqs. In the future, we'll get it from policy
+func GetStrategy() ScoreStrategy {
+	return uniformStrategy
+}
+
+// calculates the final pairing score for all slot groups (with strategy)
+// we calculate only the diff between the current and previous slot groups
+func CalcPairingScore(scores []*PairingScore, strategy ScoreStrategy, diffSlot *PairingSlot, minStake sdk.Int) error {
+	// calculate the score for each req for each provider
+	for _, req := range diffSlot.Reqs {
+		reqName := req.GetName()
+		weight, ok := strategy[reqName]
+		if !ok {
+			return utils.LavaFormatError("req not in strategy", sdkerrors.ErrKeyNotFound,
+				utils.Attribute{Key: "req_name", Value: reqName})
+		}
+
+		for _, score := range scores {
+			// normalize stake so we won't overflow the score result (uint64)
+			newScoreComp := req.Score(*score.Provider, weight)
+
+			// divide by previous score component (if exists) and multiply by new score
+			prevReqScoreComp, ok := score.ScoreComponents[reqName]
+			if ok {
+				if prevReqScoreComp == 0 {
+					return utils.LavaFormatError("previous req score is zero", fmt.Errorf("invalid req score"),
+						utils.Attribute{Key: "req_name", Value: reqName})
+				}
+				score.Score /= prevReqScoreComp
+			}
+			score.Score *= newScoreComp
+
+			// update the score component map
+			score.ScoreComponents[reqName] = newScoreComp
+		}
+	}
+
+	return nil
+}
+
+// given a list of scores, pick a <group-count> providers with a pseudo-random weighted choice
+func PickProviders(ctx sdk.Context, projectIndex string, scores []*PairingScore, groupCount uint64, block uint64, chainID string, epochHash []byte, indexToSkipPtr *map[int]bool) (returnedProviders []epochstoragetypes.StakeEntry) {
+	scoreSum := sdk.NewUint(0)
+	hashData := make([]byte, 0)
+	for _, providerScore := range scores {
+		scoreSum = scoreSum.Add(sdk.NewUint(providerScore.Score))
+	}
+	if scoreSum.IsZero() {
```

Comment:
if this intends to check for empty list - then better test directly (and before the above loop).
otherwise, is it possible to get a zero score sum?  (if not, treat it like the unexpected score zero in line 162)

Reply:
done. see https://github.com/lavanet/lava/commit/d30134e7c735b1c043ac5b7455dca1dff4ad612d
note that I use a normal panic because my branch doesn't have LavaFormatPanic. I've opened a task to change the panic instances after everything is done